File: android/src/main/kotlin/com/hellocuriosity/drappula/ui/createsequence/CreateSequenceViewModel.kt
package com.hellocuriosity.drappula.ui.createsequence

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hellocuriosity.drappula.ApplicationComponent
import com.hellocuriosity.drappula.CoroutineDispatchers
import com.hellocuriosity.drappula.data.SoundSequenceRepository
import com.hellocuriosity.drappula.models.Sound
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

abstract class CreateSequenceViewModel : ViewModel() {
    data class State(
        val selectedSounds: List<Sound> = emptyList(),
        val sequenceName: String = "",
        val isSaving: Boolean = false,
        val savedSuccessfully: Boolean = false,
        val error: Throwable? = null,
    )

    abstract val state: StateFlow<State>

    abstract fun addSound(sound: Sound)

    abstract fun removeSound(index: Int)

    abstract fun setName(name: String)

    abstract fun save()

    abstract fun reset()

    companion object {
        fun viewModel(component: ApplicationComponent): CreateSequenceViewModel =
            DefaultCreateSequenceViewModel(
                soundSequenceRepository = component.soundSequenceRepository,
                dispatchers = component.dispatchers,
            )
    }
}

class DefaultCreateSequenceViewModel(
    private val soundSequenceRepository: SoundSequenceRepository,
    private val dispatchers: CoroutineDispatchers,
) : CreateSequenceViewModel() {
    private val _state = MutableStateFlow(State())
    override val state: StateFlow<State> = _state.asStateFlow()

    override fun addSound(sound: Sound) {
        _state.update { it.copy(selectedSounds = it.selectedSounds + sound) }
    }

    override fun removeSound(index: Int) {
        _state.update {
            val mutable = it.selectedSounds.toMutableList()
            if (index in mutable.indices) {
                mutable.removeAt(index)
            }
            it.copy(selectedSounds = mutable)
        }
    }

    override fun setName(name: String) {
        _state.update { it.copy(sequenceName = name) }
    }

    @Suppress("TooGenericExceptionCaught")
    override fun save() {
        val currentState = _state.value
        if (currentState.sequenceName.isBlank() || currentState.selectedSounds.isEmpty()) return

        viewModelScope.launch(dispatchers.io) {
            _state.update { it.copy(isSaving = true, error = null) }
            try {
                soundSequenceRepository.create(
                    name = currentState.sequenceName.trim(),
                    sounds = currentState.selectedSounds,
                )
                _state.update { it.copy(isSaving = false, savedSuccessfully = true) }
            } catch (e: Exception) {
                _state.update { it.copy(isSaving = false, error = e) }
            }
        }
    }

    override fun reset() {
        _state.value = State()
    }
}

File: ios/drappula/ui/createsequence/CreateSequenceViewModel.swift
import Foundation
import shared

@MainActor
class CreateSequenceViewModel: ObservableObject {
    struct State: Equatable {
        var selectedSounds: [Sound] = []
        var sequenceName: String = ""
        var isSaving: Bool = false
        var savedSuccessfully: Bool = false

        static func == (lhs: State, rhs: State) -> Bool {
            return lhs.sequenceName == rhs.sequenceName
                && lhs.isSaving == rhs.isSaving
                && lhs.savedSuccessfully == rhs.savedSuccessfully
                && lhs.selectedSounds.count == rhs.selectedSounds.count
        }
    }

    @Published private(set) var state = State()

    private let repository: SoundSequenceRepository

    init(repository: SoundSequenceRepository) {
        self.repository = repository
    }

    func addSound(_ sound: Sound) {
        state.selectedSounds.append(sound)
    }

    func removeSound(at index: Int) {
        guard index >= 0 && index < state.selectedSounds.count else { return }
        state.selectedSounds.remove(at: index)
    }

    func setName(_ name: String) {
        state.sequenceName = name
    }

    func save() {
        guard !state.sequenceName.trimmingCharacters(in: .whitespaces).isEmpty,
              !state.selectedSounds.isEmpty else { return }

        state.isSaving = true
        repository.create(
            name: state.sequenceName.trimmingCharacters(in: .whitespaces),
            sounds: state.selectedSounds
        )
        state.isSaving = false
        state.savedSuccessfully = true
    }

    func reset() {
        state = State()
    }
}
